File: container-with-most-water/container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */

function maxArea(height) {
    let maxWaterHeight = 0;
    let leftIndex = 0, rightIndex = height.length - 1;

    while(leftIndex < rightIndex){
        maxWaterHeight = Math.max(maxWaterHeight,(rightIndex- leftIndex) * Math.min(height[leftIndex], height[rightIndex]));

        if (height[leftIndex] <= height[rightIndex])
            leftIndex++;
        else
            rightIndex--;
    }

    return maxWaterHeight;
}

console.log(maxArea([1,8,6,2,5,4,8,3,7]));

File: container-with-most-water/container-with-most-water.dart
import 'dart:math';

void main() {
  List<int> heights = [1, 8, 6, 2, 5, 4, 8, 3, 7];
  print(maxArea(heights));
}

int maxArea(List<int> heights) {
  int maxWaterHeight = 0;
  int leftIndex = 0;
  int rightIndex = heights.length - 1;

  int waterHeight = 0;
  while (leftIndex < rightIndex) {
    waterHeight =
        min(heights[leftIndex], heights[rightIndex]) * (rightIndex - leftIndex);
    maxWaterHeight = max(maxWaterHeight, waterHeight);
    if (heights[leftIndex] < heights[rightIndex]) {
      leftIndex++;
    } else {
      rightIndex--;
    }
  }
  return maxWaterHeight;
}
